zebra-rpc/src/server/http_request_compatibility.rs

```diff
@@ -103,11 +103,36 @@ impl FixHttpRequestMiddleware {
             .replace(", \"jsonrpc\": \"1.0\"", "")
     }
 
-    /// If the `content-type` HTTP header is not present,
-    /// add an `application/json` content type header.
-    pub fn add_missing_content_type_header(headers: &mut hyper::header::HeaderMap) {
-        headers
-            .entry(hyper::header::CONTENT_TYPE)
-            .or_insert(hyper::header::HeaderValue::from_static("application/json"));
+    /// Insert or replace client supplied `content-type` HTTP header to `application/json` in the following cases:
+    ///
+    /// - no `content-type` supplied.
+    /// - supplied `content-type` is `text/plain`.
+    ///
+    /// `application/json` is the only `content-type` accepted by the Zebra rpc endpoint:
+    ///
+    /// <https://github.com/paritytech/jsonrpc/blob/38af3c9439aa75481805edf6c05c6622a5ab1e70/http/src/handler.rs#L582-L584>
+    pub fn insert_or_replace_content_type_header(headers: &mut hyper::header::HeaderMap) {
+        match headers.entry(hyper::header::CONTENT_TYPE) {
+            hyper::header::Entry::Vacant(_) => {
+                headers.insert(
+                    hyper::header::CONTENT_TYPE,
+                    hyper::header::HeaderValue::from_static("application/json"),
+                );
+            }
+            hyper::header::Entry::Occupied(header_data) => {
+                if header_data
+                    .iter()
+                    .filter_map(|x| x.to_str().ok())
+                    .map(|s| s == "text/plain")
+                    .next()
+                    .unwrap_or(false)
+                {
+                    headers.insert(
+                        hyper::header::CONTENT_TYPE,
+                        hyper::header::HeaderValue::from_static("application/json"),
+                    );
+                }
```

Comment:
There is probably a simpler way to do this but i was not able to find it.

Reply:
Thanks, that simplifies the method code a lot. Applied at https://github.com/ZcashFoundation/zebra/pull/6885/commits/5d6a55703b16357f598966dfdc87cc729b5fb3a7